scipy/linalg/_decomp_cossin.py

```diff
@@ -0,0 +1,211 @@
+    else:
+        if len(X) != 4:
+            raise ValueError("When p and q are None, exactly four arrays"
+                             " should be in X, got {}".format(len(X)))
+
+        x11, x12, x21, x22 = [np.atleast_2d(x) for x in X]
+        for name, block in zip(["x11", "x12", "x21", "x22"],
+                               [x11, x12, x21, x22]):
+            if block.shape[1] == 0:
+                raise ValueError("{} can't be empty".format(name))
+        p, q = x11.shape
+        mmp, mmq = x22.shape
+
+        if x12.shape != (p, mmq):
+            raise ValueError("Invalid x12 dimensions: desired {}, "
+                             "got {}".format((p, mmq), x12.shape))
+
+        if x21.shape != (mmp, q):
+            raise ValueError("Invalid x21 dimensions: desired {}, "
+                             "got {}".format((mmp, q), x21.shape))
+
+        if p + mmp != q + mmq:
+            raise ValueError("The subblocks have compatible sizes but "
+                             "don't form a square array (instead they form a"
+                             " {}x{} array). This might be due to missing "
+                             "p, q arguments.".format(p + mmp, q + mmq))
+
+        m = p + mmp
+
+    cplx = any([np.iscomplexobj(x) for x in [x11, x12, x21, x22]])
+    driver = "uncsd" if cplx else "orcsd"
+    csd, csd_lwork = get_lapack_funcs([driver, driver + "_lwork"],
+                                      [x11, x12, x21, x22])
+    lwork = _compute_lwork(csd_lwork, m=m, p=p, q=q)
+    lwork_args = ({'lwork': lwork[0], 'lrwork': lwork[1]} if cplx else
+                  {'lwork': lwork})
+    *_, theta, u1, u2, v1h, v2h, info = csd(x11=x11, x12=x12, x21=x21, x22=x22,
+                                            compute_u1=compute_u,
+                                            compute_u2=compute_u,
+                                            compute_v1t=compute_vh,
+                                            compute_v2t=compute_vh,
+                                            trans=False, signs=swap_sign,
+                                            **lwork_args)
+    if separate:
+        return (u1, u2), theta, (v1h, v2h)
+    method_name = csd.typecode + driver
+    if info < 0:
+        raise ValueError('illegal value in argument {} of internal {}'
+                         .format(-info, method_name))
+    if info > 0:
+        raise LinAlgError("{} did not converge: {}".format(method_name, info))
+
+    U = block_diag(u1, u2)
+    VDH = block_diag(v1h, v2h)
+
+    # Construct the middle factor CS
+    c = np.diag(np.cos(theta))
+    s = np.diag(np.sin(theta))
+    r = min(p, q, m - p, m - q)
+    n11 = min(p, q) - r
+    n12 = min(p, m - q) - r
+    n21 = min(m - p, q) - r
+    n22 = min(m - p, m - q) - r
+    Id = np.eye(np.max([n11, n12, n21, n22, r]), dtype=theta.dtype)
+    CS = np.zeros((m, m), dtype=theta.dtype)
+
+    CS[:n11, :n11] = Id[:n11, :n11]
+    CS[n11 + r:
```

Comment:
I would say let's tidy this part a bit via 
```
xs, ys = n11 +r, n11 + n21 + n22 + 2 * r
xe, ye = xs + n12, ys + n12
CS[xs:xe, ys:ye] = Id[:n12, :n12] if swap_sign else -Id[:n12, :n12]
```
otherwise it's really difficult to read. 4 variable assignments shouldn't cause any performance problems.